{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 336,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd             \n",
    "import numpy as np             \n",
    "import matplotlib.pyplot as plt  \n",
    "import seaborn as sns           \n",
    "\n",
    "from sklearn.preprocessing import OneHotEncoder # 원핫인코딩\n",
    "from sklearn.preprocessing import LabelEncoder # 라벨인코딩\n",
    "from sklearn.model_selection import train_test_split  # 데이터 분할\n",
    "from sklearn.preprocessing import StandardScaler      # 스케일링\n",
    "from sklearn.impute import SimpleImputer              # 결측치 처리\n",
    "from sklearn.ensemble import RandomForestClassifier   # 분류 모델\n",
    "from sklearn.linear_model import LogisticRegression   # 분류 모델\n",
    "from sklearn.metrics import accuracy_score, confusion_matrix, classification_report  # 모델 평가\n",
    "from sklearn.linear_model import LinearRegression     # 회귀 모델\n",
    "from sklearn.ensemble import RandomForestRegressor   # 회귀 모델\n",
    "from sklearn.metrics import mean_absolute_error, mean_squared_error, r2_score  # 모델 평가\n",
    "\n",
    "# 경고 메시지 무시\n",
    "import warnings\n",
    "warnings.filterwarnings('ignore')\n",
    "\n",
    "## 한글 깨짐 방지\n",
    "#pip install koreanize_matplotlib\n",
    "import koreanize_matplotlib "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# sampleMap 데이터"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 337,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(1218, 205)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>sampleID</th>\n",
       "      <th>AJCC_Stage_nature2012</th>\n",
       "      <th>Age_at_Initial_Pathologic_Diagnosis_nature2012</th>\n",
       "      <th>CN_Clusters_nature2012</th>\n",
       "      <th>Converted_Stage_nature2012</th>\n",
       "      <th>Days_to_Date_of_Last_Contact_nature2012</th>\n",
       "      <th>Days_to_date_of_Death_nature2012</th>\n",
       "      <th>ER_Status_nature2012</th>\n",
       "      <th>Gender_nature2012</th>\n",
       "      <th>...</th>\n",
       "      <th>pan.samplesID</th>\n",
       "      <th>cancer.type</th>\n",
       "      <th>Subtype_mRNA</th>\n",
       "      <th>Subtype_DNAmeth</th>\n",
       "      <th>Subtype_protein</th>\n",
       "      <th>Subtype_miRNA</th>\n",
       "      <th>Subtype_CNA</th>\n",
       "      <th>Subtype_Integrative</th>\n",
       "      <th>Subtype_other</th>\n",
       "      <th>Subtype_Selected</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>TCGA-3C-AAAU-01</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>TCGA-3C-AAAU-01</td>\n",
       "      <td>BRCA</td>\n",
       "      <td>LumA</td>\n",
       "      <td>NA_character_</td>\n",
       "      <td>-2147483648</td>\n",
       "      <td>NA_character_</td>\n",
       "      <td>NA_character_</td>\n",
       "      <td>NA_character_</td>\n",
       "      <td>NA_character_</td>\n",
       "      <td>BRCA.LumA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>TCGA-3C-AALI-01</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>TCGA-3C-AALI-01</td>\n",
       "      <td>BRCA</td>\n",
       "      <td>Her2</td>\n",
       "      <td>NA_character_</td>\n",
       "      <td>-2147483648</td>\n",
       "      <td>NA_character_</td>\n",
       "      <td>NA_character_</td>\n",
       "      <td>NA_character_</td>\n",
       "      <td>NA_character_</td>\n",
       "      <td>BRCA.Her2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>TCGA-3C-AALJ-01</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>TCGA-3C-AALJ-01</td>\n",
       "      <td>BRCA</td>\n",
       "      <td>LumB</td>\n",
       "      <td>NA_character_</td>\n",
       "      <td>-2147483648</td>\n",
       "      <td>NA_character_</td>\n",
       "      <td>NA_character_</td>\n",
       "      <td>NA_character_</td>\n",
       "      <td>NA_character_</td>\n",
       "      <td>BRCA.LumB</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>TCGA-3C-AALK-01</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>TCGA-3C-AALK-01</td>\n",
       "      <td>BRCA</td>\n",
       "      <td>LumA</td>\n",
       "      <td>NA_character_</td>\n",
       "      <td>-2147483648</td>\n",
       "      <td>NA_character_</td>\n",
       "      <td>NA_character_</td>\n",
       "      <td>NA_character_</td>\n",
       "      <td>NA_character_</td>\n",
       "      <td>BRCA.LumA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>TCGA-4H-AAAK-01</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>TCGA-4H-AAAK-01</td>\n",
       "      <td>BRCA</td>\n",
       "      <td>LumA</td>\n",
       "      <td>NA_character_</td>\n",
       "      <td>-2147483648</td>\n",
       "      <td>NA_character_</td>\n",
       "      <td>NA_character_</td>\n",
       "      <td>NA_character_</td>\n",
       "      <td>NA_character_</td>\n",
       "      <td>BRCA.LumA</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 205 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   Unnamed: 0         sampleID AJCC_Stage_nature2012  \\\n",
       "0           0  TCGA-3C-AAAU-01                   NaN   \n",
       "1           1  TCGA-3C-AALI-01                   NaN   \n",
       "2           2  TCGA-3C-AALJ-01                   NaN   \n",
       "3           3  TCGA-3C-AALK-01                   NaN   \n",
       "4           4  TCGA-4H-AAAK-01                   NaN   \n",
       "\n",
       "   Age_at_Initial_Pathologic_Diagnosis_nature2012  CN_Clusters_nature2012  \\\n",
       "0                                             NaN                     NaN   \n",
       "1                                             NaN                     NaN   \n",
       "2                                             NaN                     NaN   \n",
       "3                                             NaN                     NaN   \n",
       "4                                             NaN                     NaN   \n",
       "\n",
       "  Converted_Stage_nature2012  Days_to_Date_of_Last_Contact_nature2012  \\\n",
       "0                        NaN                                      NaN   \n",
       "1                        NaN                                      NaN   \n",
       "2                        NaN                                      NaN   \n",
       "3                        NaN                                      NaN   \n",
       "4                        NaN                                      NaN   \n",
       "\n",
       "   Days_to_date_of_Death_nature2012 ER_Status_nature2012 Gender_nature2012  \\\n",
       "0                               NaN                  NaN               NaN   \n",
       "1                               NaN                  NaN               NaN   \n",
       "2                               NaN                  NaN               NaN   \n",
       "3                               NaN                  NaN               NaN   \n",
       "4                               NaN                  NaN               NaN   \n",
       "\n",
       "   ...    pan.samplesID  cancer.type  Subtype_mRNA  Subtype_DNAmeth  \\\n",
       "0  ...  TCGA-3C-AAAU-01         BRCA          LumA    NA_character_   \n",
       "1  ...  TCGA-3C-AALI-01         BRCA          Her2    NA_character_   \n",
       "2  ...  TCGA-3C-AALJ-01         BRCA          LumB    NA_character_   \n",
       "3  ...  TCGA-3C-AALK-01         BRCA          LumA    NA_character_   \n",
       "4  ...  TCGA-4H-AAAK-01         BRCA          LumA    NA_character_   \n",
       "\n",
       "  Subtype_protein  Subtype_miRNA    Subtype_CNA Subtype_Integrative  \\\n",
       "0     -2147483648  NA_character_  NA_character_       NA_character_   \n",
       "1     -2147483648  NA_character_  NA_character_       NA_character_   \n",
       "2     -2147483648  NA_character_  NA_character_       NA_character_   \n",
       "3     -2147483648  NA_character_  NA_character_       NA_character_   \n",
       "4     -2147483648  NA_character_  NA_character_       NA_character_   \n",
       "\n",
       "   Subtype_other  Subtype_Selected  \n",
       "0  NA_character_         BRCA.LumA  \n",
       "1  NA_character_         BRCA.Her2  \n",
       "2  NA_character_         BRCA.LumB  \n",
       "3  NA_character_         BRCA.LumA  \n",
       "4  NA_character_         BRCA.LumA  \n",
       "\n",
       "[5 rows x 205 columns]"
      ]
     },
     "execution_count": 337,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sampleMap = pd.read_csv('/Users/zoohunn/Desktop/비어플/비어플[의료]/tcga/sampleMap_BRCA_clinicalMatrix.csv', sep=',')\n",
    "print(sampleMap.shape)\n",
    "sampleMap.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 338,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "sampleID                                             0\n",
       "AJCC_Stage_nature2012                              439\n",
       "Age_at_Initial_Pathologic_Diagnosis_nature2012     285\n",
       "CN_Clusters_nature2012                             445\n",
       "Converted_Stage_nature2012                         427\n",
       "                                                  ... \n",
       "Subtype_miRNA                                     1218\n",
       "Subtype_CNA                                       1218\n",
       "Subtype_Integrative                               1218\n",
       "Subtype_other                                     1218\n",
       "Subtype_Selected                                     0\n",
       "Length: 204, dtype: int64"
      ]
     },
     "execution_count": 338,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sampleMap = sampleMap.drop('Unnamed: 0', axis=1)\n",
    "sampleMap.replace('NA_character_', np.nan, inplace=True)\n",
    "sampleMap.replace(-2147483648, np.nan, inplace=True)\n",
    "sampleMap.isnull().sum()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "sampleMape 데이터는 (1218, 205)의 구조며, 많은 열에 결측치가 과도하게 존재   \n",
    "결측치가 대부분인 열을 우선 제거하고, 생존 분석과 직접 관련이 없는 변수들을 유형별로 묶어 삭제   \n",
    "변수 하나하나를 모두 설명하지 않고, 생존 분석과 무관한 변수들을 유형별로 구분하여 삭제하는 방향으로 진행하겠음"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "생존 분석 무관 변수 유형:   \n",
    "\t•\t유전자 클러스터 변수   \n",
    "\t•\t행정적 메타 정보 변수   \n",
    "\t•\t연구 목적 변수   \n",
    "\t•\t이차 정보(중복) 변수   \n",
    "\t•   GENOMIC_ID_TCGA_BRCA_로 시작하는 변수\n",
    "\t•   조직(tissue) 변수"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 339,
   "metadata": {},
   "outputs": [],
   "source": [
    "## 유전자 클러스터 변수 제거\n",
    "# 고차원 데이터로 너무 많은 변수가 존재하면 모델이 복잡해지고, 과적합이 발생할 수 있음\n",
    "# 생존 분석의 중요한 명확하고 직관적인 변수가 아닌, 클러스터 변수는 의미가 모호\n",
    "\n",
    "# 유전자 클러스터 변수 목록\n",
    "genomic_cluster_columns = [\n",
    "    'Integrated_Clusters_no_exp__nature2012',\n",
    "    'Integrated_Clusters_unsup_exp__nature2012',\n",
    "    'Integrated_Clusters_with_PAM50__nature2012',\n",
    "    'CN_Clusters_nature2012',\n",
    "    'methylation_Clusters_nature2012',\n",
    "    'miRNA_Clusters_nature2012',\n",
    "    'RPPA_Clusters_nature2012',\n",
    "    'SigClust_Intrinsic_mRNA_nature2012',\n",
    "    'SigClust_Unsupervised_mRNA_nature2012'\n",
    "]\n",
    "\n",
    "## 연구 목적 변수 제거\n",
    "# 연구 목적 변수들은 특정 프로젝트나 연구에서 데이터 관리를 위해 추가된 변수\n",
    "# 연구 목적 변수들은 이러한 생존 시간과 직접적인 관련이 없으므로 불필요\n",
    "\n",
    "# 연구 목적 변수 목록\n",
    "research_columns = [\n",
    "    'Survival_Data_Form_nature2012',\n",
    "    '_PANCAN_CNA_PANCAN_K8',\n",
    "    '_PANCAN_Cluster_Cluster_PANCAN',\n",
    "    '_PANCAN_DNAMethyl_BRCA',\n",
    "    '_PANCAN_DNAMethyl_PANCAN',\n",
    "    '_PANCAN_RPPA_PANCAN_K8',\n",
    "    '_PANCAN_UNC_RNAseq_PANCAN_K16',\n",
    "    '_PANCAN_miRNA_PANCAN',\n",
    "    '_PANCAN_mirna_BRCA',\n",
    "    '_PANCAN_mutation_PANCAN',\n",
    "    '_PATIENT',\n",
    "    '_cohort',\n",
    "    '_primary_disease',\n",
    "    '_primary_site'\n",
    "]\n",
    "\n",
    "## 행정적 메타 변수 제거\n",
    "# 데이터 관리 및 식별을 위해 추가된 변수\n",
    "# 환자 식별자, 코호트 정보, 연구 프로젝트 정보 등을 포함(중복됨)\n",
    "\n",
    "# 행정적 메타 정보 변수 목록\n",
    "admin_columns = [\n",
    "    'bcr_patient_barcode', 'bcr_sample_barcode', 'bcr_followup_barcode', \n",
    "    '_cohort', '_primary_disease', '_primary_site', '_INTEGRATION', '_PATIENT', \n",
    "    'Redaction', 'pan.samplesID'\n",
    "]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 340,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "제거된 변수 개수: 32\n",
      "제거된 변수 목록:\n",
      "- Integrated_Clusters_no_exp__nature2012\n",
      "- Integrated_Clusters_unsup_exp__nature2012\n",
      "- Integrated_Clusters_with_PAM50__nature2012\n",
      "- CN_Clusters_nature2012\n",
      "- methylation_Clusters_nature2012\n",
      "- miRNA_Clusters_nature2012\n",
      "- RPPA_Clusters_nature2012\n",
      "- SigClust_Intrinsic_mRNA_nature2012\n",
      "- SigClust_Unsupervised_mRNA_nature2012\n",
      "- Survival_Data_Form_nature2012\n",
      "- _PANCAN_CNA_PANCAN_K8\n",
      "- _PANCAN_Cluster_Cluster_PANCAN\n",
      "- _PANCAN_DNAMethyl_BRCA\n",
      "- _PANCAN_DNAMethyl_PANCAN\n",
      "- _PANCAN_RPPA_PANCAN_K8\n",
      "- _PANCAN_UNC_RNAseq_PANCAN_K16\n",
      "- _PANCAN_miRNA_PANCAN\n",
      "- _PANCAN_mirna_BRCA\n",
      "- _PANCAN_mutation_PANCAN\n",
      "- _PATIENT\n",
      "- _cohort\n",
      "- _primary_disease\n",
      "- _primary_site\n",
      "- bcr_patient_barcode\n",
      "- bcr_sample_barcode\n",
      "- bcr_followup_barcode\n",
      "- _cohort\n",
      "- _primary_disease\n",
      "- _primary_site\n",
      "- _INTEGRATION\n",
      "- _PATIENT\n",
      "- pan.samplesID\n",
      "제거 후 데이터 형태: (1218, 176)\n"
     ]
    }
   ],
   "source": [
    "# 제거할 변수 목록 합치기\n",
    "remove_columns = genomic_cluster_columns + research_columns + admin_columns\n",
    "\n",
    "# 실제 데이터에 존재하는 변수만 필터링\n",
    "remove_columns = [col for col in remove_columns if col in sampleMap.columns]\n",
    "\n",
    "# 불필요 변수 제거\n",
    "sampleMap2 = sampleMap.drop(columns=remove_columns)\n",
    "\n",
    "# 제거 완료 확인\n",
    "print(f\"제거된 변수 개수: {len(remove_columns)}\")\n",
    "print(\"제거된 변수 목록:\")\n",
    "for col in remove_columns:\n",
    "    print(f\"- {col}\")\n",
    "\n",
    "# 제거 후 데이터 확인\n",
    "print(\"제거 후 데이터 형태:\", sampleMap2.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 341,
   "metadata": {},
   "outputs": [],
   "source": [
    "## 어느정도 결측치의 비율이 있는 데이터를 제거해야할지 확신이 안서지만 70% 이상의 결측치가 있는 열을 제거해보겠음,,\n",
    "\n",
    "# 결측치 비율 계산\n",
    "missing_percentage = sampleMap2.isnull().mean() * 100\n",
    "\n",
    "# 70% 이상 결측치인 열 목록\n",
    "high_missing_columns = missing_percentage[missing_percentage > 70].index.tolist()\n",
    "\n",
    "sampleMap2 = sampleMap2.drop(columns=high_missing_columns)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 342,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "제거된 '_GENOMIC_ID_TCGA_BRCA'로 시작하는 변수 개수: 18\n",
      "제거된 변수 목록:\n",
      "- _GENOMIC_ID_TCGA_BRCA_exp_HiSeqV2_exon\n",
      "- _GENOMIC_ID_TCGA_BRCA_exp_HiSeqV2_PANCAN\n",
      "- _GENOMIC_ID_TCGA_BRCA_RPPA_RBN\n",
      "- _GENOMIC_ID_TCGA_BRCA_mutation\n",
      "- _GENOMIC_ID_TCGA_BRCA_PDMRNAseq\n",
      "- _GENOMIC_ID_TCGA_BRCA_hMethyl450\n",
      "- _GENOMIC_ID_TCGA_BRCA_RPPA\n",
      "- _GENOMIC_ID_TCGA_BRCA_PDMRNAseqCNV\n",
      "- _GENOMIC_ID_TCGA_BRCA_mutation_curated_wustl_gene\n",
      "- _GENOMIC_ID_TCGA_BRCA_PDMarrayCNV\n",
      "- _GENOMIC_ID_TCGA_BRCA_miRNA_HiSeq\n",
      "- _GENOMIC_ID_TCGA_BRCA_mutation_wustl_gene\n",
      "- _GENOMIC_ID_TCGA_BRCA_exp_HiSeqV2_percentile\n",
      "- _GENOMIC_ID_TCGA_BRCA_gistic2thd\n",
      "- _GENOMIC_ID_TCGA_BRCA_G4502A_07_3\n",
      "- _GENOMIC_ID_TCGA_BRCA_exp_HiSeqV2\n",
      "- _GENOMIC_ID_TCGA_BRCA_gistic2\n",
      "- _GENOMIC_ID_TCGA_BRCA_PDMarray\n"
     ]
    }
   ],
   "source": [
    "## GENOMIC_ID_TCGA_BRCA_ 시작하는 변수들 제거\n",
    "# 주로 유전자 발현, 유전자 복제 수 변이, DNA 메틸화 등을 포함\n",
    "# 생존 분석의 관점에서 보면 대부분 직접적으로 필요하지 않음\n",
    "\n",
    "genomic_id_columns = [col for col in sampleMap2.columns if col.startswith('_GENOMIC_ID_TCGA_BRCA')]\n",
    "\n",
    "\n",
    "sampleMap2 = sampleMap2.drop(columns=genomic_id_columns)\n",
    "print(f\"제거된 '_GENOMIC_ID_TCGA_BRCA'로 시작하는 변수 개수: {len(genomic_id_columns)}\")\n",
    "print(\"제거된 변수 목록:\")\n",
    "for col in genomic_id_columns:\n",
    "    print(f\"- {col}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 343,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "'tissue' 관련 변수 개수: 4\n",
      "tissue 관련 변수 목록:\n",
      "- tissue_prospective_collection_indicator\n",
      "- tissue_retrospective_collection_indicator\n",
      "- tissue_source_site\n",
      "- tumor_tissue_site\n"
     ]
    }
   ],
   "source": [
    "## tissue 변수 제거\n",
    "\n",
    "# 'tissue'라는 단어가 포함된 모든 변수명을 필터링\n",
    "tissue_columns = [col for col in sampleMap2.columns if 'tissue' in col]\n",
    "\n",
    "# tissue 관련 변수들 목록 출력\n",
    "print(f\"'tissue' 관련 변수 개수: {len(tissue_columns)}\")\n",
    "print(\"tissue 관련 변수 목록:\")\n",
    "for col in tissue_columns:\n",
    "    print(f\"- {col}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 344,
   "metadata": {},
   "outputs": [],
   "source": [
    "# 생존 분석 관점에서 'tissue' 변수는 필요하지 않다고 판단하여 제거\n",
    "sampleMap2 = sampleMap2.drop(columns=tissue_columns)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 345,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "AJCC_Stage_nature2012  Converted_Stage_nature2012\n",
      "Stage IIA              Stage IIA                     209\n",
      "Stage IIB              Stage IIB                      95\n",
      "Stage IIIA             Stage IIIA                     67\n",
      "Stage IA               Stage I                        60\n",
      "Stage I                Stage I                        56\n",
      "Stage IIB              No_Conversion                  53\n",
      "Stage IIIA             No_Conversion                  32\n",
      "Stage IIA              No_Conversion                  31\n",
      "Stage IIIC             Stage IIIC                     22\n",
      "Stage II               Stage IIA                      18\n",
      "                       No_Conversion                  11\n",
      "Stage III              Stage IIIA                     11\n",
      "Stage II               Stage IIB                      11\n",
      "Stage IV               No_Conversion                  11\n",
      "Stage IIIC             No_Conversion                   9\n",
      "Stage IIIB             Stage IIIB                      9\n",
      "                       No_Conversion                   8\n",
      "Stage IB               No_Conversion                   6\n",
      "Stage IIIB             Stage IIIC                      5\n",
      "Stage III              No_Conversion                   5\n",
      "Stage IIB              Stage IIA                       5\n",
      "Stage IIA              Stage IIB                       5\n",
      "Stage IIB              Stage IIIA                      4\n",
      "Stage X                Stage IIA                       4\n",
      "Stage IV               Stage IV                        4\n",
      "Stage IB               Stage I                         3\n",
      "Stage IIIA             Stage IIA                       3\n",
      "Stage IA               No_Conversion                   3\n",
      "Stage IIA              Stage I                         3\n",
      "Stage I                Stage IIA                       2\n",
      "Stage IB               Stage IIB                       2\n",
      "Stage III              Stage IIIC                      2\n",
      "Stage X                Stage IIIC                      2\n",
      "Stage IIIC             Stage IIIA                      1\n",
      "Stage IA               Stage IIA                       1\n",
      "Stage IIIB             Stage                           1\n",
      "Stage IIA              Stage IIIA                      1\n",
      "Stage X                No_Conversion                   1\n",
      "                       Stage I                         1\n",
      "                       Stage IIIA                      1\n",
      "                       Stage IIIB                      1\n",
      "Name: count, dtype: int64\n"
     ]
    }
   ],
   "source": [
    "## 두 변수 의미가 중복돼 하나 제거(암 병기 변수)\n",
    "print(sampleMap2[['AJCC_Stage_nature2012', 'Converted_Stage_nature2012']].value_counts())\n",
    "sampleMap2 = sampleMap2.drop(columns=['Converted_Stage_nature2012'])\n",
    "cleaned_sampleMap_df = sampleMap2.drop(columns=['_GENOMIC_ID_data/public/TCGA/BRCA/miRNA_HiSeq_gene', 'cancer.type','Subtype_mRNA'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 346,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(sampleID                                            0\n",
       " AJCC_Stage_nature2012                             439\n",
       " Age_at_Initial_Pathologic_Diagnosis_nature2012    285\n",
       " Days_to_Date_of_Last_Contact_nature2012           351\n",
       " ER_Status_nature2012                              436\n",
       "                                                  ... \n",
       " targeted_molecular_therapy                        583\n",
       " vial_number                                       192\n",
       " vital_status                                        3\n",
       " year_of_initial_pathologic_diagnosis                5\n",
       " Subtype_Selected                                    0\n",
       " Length: 75, dtype: int64,\n",
       " (1218, 75))"
      ]
     },
     "execution_count": 346,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cleaned_sampleMap_df.isnull().sum(), cleaned_sampleMap_df.shape"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Survival 데이터"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 347,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(1215, 22)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>sample</th>\n",
       "      <th>_PATIENT</th>\n",
       "      <th>OS</th>\n",
       "      <th>OS.time</th>\n",
       "      <th>DSS</th>\n",
       "      <th>DSS.time</th>\n",
       "      <th>DFI</th>\n",
       "      <th>DFI.time</th>\n",
       "      <th>PFI</th>\n",
       "      <th>...</th>\n",
       "      <th>pan.samplesID</th>\n",
       "      <th>cancer.type</th>\n",
       "      <th>Subtype_mRNA</th>\n",
       "      <th>Subtype_DNAmeth</th>\n",
       "      <th>Subtype_protein</th>\n",
       "      <th>Subtype_miRNA</th>\n",
       "      <th>Subtype_CNA</th>\n",
       "      <th>Subtype_Integrative</th>\n",
       "      <th>Subtype_other</th>\n",
       "      <th>Subtype_Selected</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>TCGA-3C-AAAU-01</td>\n",
       "      <td>TCGA-3C-AAAU</td>\n",
       "      <td>0</td>\n",
       "      <td>4047.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4047.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1808.0</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>TCGA-3C-AAAU-01</td>\n",
       "      <td>BRCA</td>\n",
       "      <td>LumA</td>\n",
       "      <td>NA_character_</td>\n",
       "      <td>-2147483648</td>\n",
       "      <td>NA_character_</td>\n",
       "      <td>NA_character_</td>\n",
       "      <td>NA_character_</td>\n",
       "      <td>NA_character_</td>\n",
       "      <td>BRCA.LumA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>TCGA-3C-AALI-01</td>\n",
       "      <td>TCGA-3C-AALI</td>\n",
       "      <td>0</td>\n",
       "      <td>4005.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4005.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4005.0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>TCGA-3C-AALI-01</td>\n",
       "      <td>BRCA</td>\n",
       "      <td>Her2</td>\n",
       "      <td>NA_character_</td>\n",
       "      <td>-2147483648</td>\n",
       "      <td>NA_character_</td>\n",
       "      <td>NA_character_</td>\n",
       "      <td>NA_character_</td>\n",
       "      <td>NA_character_</td>\n",
       "      <td>BRCA.Her2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>TCGA-3C-AALJ-01</td>\n",
       "      <td>TCGA-3C-AALJ</td>\n",
       "      <td>0</td>\n",
       "      <td>1474.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1474.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1474.0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>TCGA-3C-AALJ-01</td>\n",
       "      <td>BRCA</td>\n",
       "      <td>LumB</td>\n",
       "      <td>NA_character_</td>\n",
       "      <td>-2147483648</td>\n",
       "      <td>NA_character_</td>\n",
       "      <td>NA_character_</td>\n",
       "      <td>NA_character_</td>\n",
       "      <td>NA_character_</td>\n",
       "      <td>BRCA.LumB</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>TCGA-3C-AALK-01</td>\n",
       "      <td>TCGA-3C-AALK</td>\n",
       "      <td>0</td>\n",
       "      <td>1448.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1448.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>TCGA-3C-AALK-01</td>\n",
       "      <td>BRCA</td>\n",
       "      <td>LumA</td>\n",
       "      <td>NA_character_</td>\n",
       "      <td>-2147483648</td>\n",
       "      <td>NA_character_</td>\n",
       "      <td>NA_character_</td>\n",
       "      <td>NA_character_</td>\n",
       "      <td>NA_character_</td>\n",
       "      <td>BRCA.LumA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>TCGA-4H-AAAK-01</td>\n",
       "      <td>TCGA-4H-AAAK</td>\n",
       "      <td>0</td>\n",
       "      <td>348.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>348.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>348.0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>TCGA-4H-AAAK-01</td>\n",
       "      <td>BRCA</td>\n",
       "      <td>LumA</td>\n",
       "      <td>NA_character_</td>\n",
       "      <td>-2147483648</td>\n",
       "      <td>NA_character_</td>\n",
       "      <td>NA_character_</td>\n",
       "      <td>NA_character_</td>\n",
       "      <td>NA_character_</td>\n",
       "      <td>BRCA.LumA</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 22 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   Unnamed: 0           sample      _PATIENT  OS  OS.time  DSS  DSS.time  DFI  \\\n",
       "0           0  TCGA-3C-AAAU-01  TCGA-3C-AAAU   0   4047.0  0.0    4047.0  1.0   \n",
       "1           1  TCGA-3C-AALI-01  TCGA-3C-AALI   0   4005.0  0.0    4005.0  0.0   \n",
       "2           2  TCGA-3C-AALJ-01  TCGA-3C-AALJ   0   1474.0  0.0    1474.0  0.0   \n",
       "3           3  TCGA-3C-AALK-01  TCGA-3C-AALK   0   1448.0  0.0    1448.0  NaN   \n",
       "4           4  TCGA-4H-AAAK-01  TCGA-4H-AAAK   0    348.0  0.0     348.0  0.0   \n",
       "\n",
       "   DFI.time  PFI  ...    pan.samplesID cancer.type Subtype_mRNA  \\\n",
       "0    1808.0    1  ...  TCGA-3C-AAAU-01        BRCA         LumA   \n",
       "1    4005.0    0  ...  TCGA-3C-AALI-01        BRCA         Her2   \n",
       "2    1474.0    0  ...  TCGA-3C-AALJ-01        BRCA         LumB   \n",
       "3       NaN    0  ...  TCGA-3C-AALK-01        BRCA         LumA   \n",
       "4     348.0    0  ...  TCGA-4H-AAAK-01        BRCA         LumA   \n",
       "\n",
       "  Subtype_DNAmeth Subtype_protein  Subtype_miRNA    Subtype_CNA  \\\n",
       "0   NA_character_     -2147483648  NA_character_  NA_character_   \n",
       "1   NA_character_     -2147483648  NA_character_  NA_character_   \n",
       "2   NA_character_     -2147483648  NA_character_  NA_character_   \n",
       "3   NA_character_     -2147483648  NA_character_  NA_character_   \n",
       "4   NA_character_     -2147483648  NA_character_  NA_character_   \n",
       "\n",
       "  Subtype_Integrative  Subtype_other Subtype_Selected  \n",
       "0       NA_character_  NA_character_        BRCA.LumA  \n",
       "1       NA_character_  NA_character_        BRCA.Her2  \n",
       "2       NA_character_  NA_character_        BRCA.LumB  \n",
       "3       NA_character_  NA_character_        BRCA.LumA  \n",
       "4       NA_character_  NA_character_        BRCA.LumA  \n",
       "\n",
       "[5 rows x 22 columns]"
      ]
     },
     "execution_count": 347,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "survival = pd.read_csv('/Users/zoohunn/Desktop/비어플/비어플[의료]/tcga/survival_BRCA_survival_(BRCA).csv')\n",
    "print(survival.shape)\n",
    "survival.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 348,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "sample                    0\n",
       "_PATIENT                  0\n",
       "OS                        0\n",
       "OS.time                   1\n",
       "DSS                      30\n",
       "DSS.time                  1\n",
       "DFI                     177\n",
       "DFI.time                178\n",
       "PFI                       0\n",
       "PFI.time                  1\n",
       "Redaction              1209\n",
       "pan.samplesID             0\n",
       "cancer.type               0\n",
       "Subtype_mRNA              0\n",
       "Subtype_DNAmeth        1215\n",
       "Subtype_protein        1215\n",
       "Subtype_miRNA          1215\n",
       "Subtype_CNA            1215\n",
       "Subtype_Integrative    1215\n",
       "Subtype_other          1215\n",
       "Subtype_Selected          0\n",
       "dtype: int64"
      ]
     },
     "execution_count": 348,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "survival = survival.drop('Unnamed: 0', axis=1)\n",
    "survival.replace('NA_character_', np.nan, inplace=True)\n",
    "survival.replace(-2147483648, np.nan, inplace=True)\n",
    "survival.isnull().sum() "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- 전체 행, 대부분이 결측치인 열들 제거\n",
    "- 중복된 의미의 변수 _PATIENT, pan.samplesID,'cancer.type','Subtype_mRNA' 제거\n",
    "- DFI, DFI.time 각각 177, 178개의 결측치가 같은 행에 동시에 존재하는 지 확인 후 제거\n",
    "- OS.time, DSS.time, PFI.time 결측치 한 개가 같은 행에 동시에 존재하는 지 확인 후 제거"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 349,
   "metadata": {},
   "outputs": [],
   "source": [
    "survival2 = survival.copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 350,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "DFI와 DFI.time이 같은 행에서 결측치인 경우: 177개\n",
      "               sample  DFI  DFI.time\n",
      "3     TCGA-3C-AALK-01  NaN       NaN\n",
      "5     TCGA-5L-AAT0-01  NaN       NaN\n",
      "6     TCGA-5L-AAT1-01  NaN       NaN\n",
      "7     TCGA-5T-A9QA-01  NaN       NaN\n",
      "16    TCGA-A1-A0SK-01  NaN       NaN\n",
      "...               ...  ...       ...\n",
      "1193  TCGA-PL-A8LY-01  NaN       NaN\n",
      "1194  TCGA-PL-A8LZ-01  NaN       NaN\n",
      "1205  TCGA-UL-AAZ6-01  NaN       NaN\n",
      "1206  TCGA-UU-A93S-01  NaN       NaN\n",
      "1213  TCGA-Z7-A8R5-01  NaN       NaN\n",
      "\n",
      "[177 rows x 3 columns]\n"
     ]
    }
   ],
   "source": [
    "# 같은 행에서 DFI와 DFI.time이 결측치인 행 필터링\n",
    "same_missing = survival2[survival2['DFI'].isnull() & survival2['DFI.time'].isnull()]\n",
    "\n",
    "# 같은 행에서 결측치인 행 개수와 목록 출력\n",
    "print(f\"DFI와 DFI.time이 같은 행에서 결측치인 경우: {same_missing.shape[0]}개\")\n",
    "print(same_missing[['sample', 'DFI', 'DFI.time']])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 351,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "제거된 행 개수: 177\n"
     ]
    }
   ],
   "source": [
    "# DFI와 DFI.time이 모두 결측치인 행 제거\n",
    "cleaned_survival_df = survival2[~(survival2['DFI'].isnull() & survival2['DFI.time'].isnull())]\n",
    "\n",
    "# 제거된 행 개수 확인\n",
    "removed_count = survival2.shape[0] - cleaned_survival_df.shape[0]\n",
    "print(f\"제거된 행 개수: {removed_count}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 352,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample</th>\n",
       "      <th>_PATIENT</th>\n",
       "      <th>OS</th>\n",
       "      <th>OS.time</th>\n",
       "      <th>DSS</th>\n",
       "      <th>DSS.time</th>\n",
       "      <th>DFI</th>\n",
       "      <th>DFI.time</th>\n",
       "      <th>PFI</th>\n",
       "      <th>PFI.time</th>\n",
       "      <th>...</th>\n",
       "      <th>pan.samplesID</th>\n",
       "      <th>cancer.type</th>\n",
       "      <th>Subtype_mRNA</th>\n",
       "      <th>Subtype_DNAmeth</th>\n",
       "      <th>Subtype_protein</th>\n",
       "      <th>Subtype_miRNA</th>\n",
       "      <th>Subtype_CNA</th>\n",
       "      <th>Subtype_Integrative</th>\n",
       "      <th>Subtype_other</th>\n",
       "      <th>Subtype_Selected</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1176</th>\n",
       "      <td>TCGA-OL-A66H-01</td>\n",
       "      <td>TCGA-OL-A66H</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>TCGA-OL-A66H-01</td>\n",
       "      <td>BRCA</td>\n",
       "      <td>LumA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>BRCA.LumA</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1 rows × 21 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "               sample      _PATIENT  OS  OS.time  DSS  DSS.time  DFI  \\\n",
       "1176  TCGA-OL-A66H-01  TCGA-OL-A66H   0      NaN  0.0       NaN  0.0   \n",
       "\n",
       "      DFI.time  PFI  PFI.time  ...    pan.samplesID cancer.type Subtype_mRNA  \\\n",
       "1176       NaN    0       NaN  ...  TCGA-OL-A66H-01        BRCA         LumA   \n",
       "\n",
       "     Subtype_DNAmeth  Subtype_protein  Subtype_miRNA  Subtype_CNA  \\\n",
       "1176             NaN              NaN            NaN          NaN   \n",
       "\n",
       "      Subtype_Integrative  Subtype_other  Subtype_Selected  \n",
       "1176                  NaN            NaN         BRCA.LumA  \n",
       "\n",
       "[1 rows x 21 columns]"
      ]
     },
     "execution_count": 352,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cleaned_survival_df[\n",
    "    cleaned_survival_df[[\"OS.time\", \"DSS.time\", \"PFI.time\"]].isnull().any(axis=1)\n",
    "]\n",
    "# 같은 행에 결측치 존재, 삭제하기"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 353,
   "metadata": {},
   "outputs": [],
   "source": [
    "cleaned_survival_df = cleaned_survival_df.drop(columns=['_PATIENT', 'Subtype_DNAmeth','Subtype_protein','Subtype_miRNA','Subtype_CNA','Subtype_Integrative','Subtype_other','Redaction','pan.samplesID','cancer.type','Subtype_mRNA'])\n",
    "cleaned_survival_df = cleaned_survival_df[cleaned_survival_df[\"sample\"] != \"TCGA-OL-A66H-01\"].reset_index(drop=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 354,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "((1037, 10),\n",
       "             sample  OS  OS.time  DSS  DSS.time  DFI  DFI.time  PFI  PFI.time  \\\n",
       " 0  TCGA-3C-AAAU-01   0   4047.0  0.0    4047.0  1.0    1808.0    1    1808.0   \n",
       " 1  TCGA-3C-AALI-01   0   4005.0  0.0    4005.0  0.0    4005.0    0    4005.0   \n",
       " 2  TCGA-3C-AALJ-01   0   1474.0  0.0    1474.0  0.0    1474.0    0    1474.0   \n",
       " 3  TCGA-4H-AAAK-01   0    348.0  0.0     348.0  0.0     348.0    0     348.0   \n",
       " 4  TCGA-A1-A0SB-01   0    259.0  0.0     259.0  0.0     259.0    0     259.0   \n",
       " \n",
       "   Subtype_Selected  \n",
       " 0        BRCA.LumA  \n",
       " 1        BRCA.Her2  \n",
       " 2        BRCA.LumB  \n",
       " 3        BRCA.LumA  \n",
       " 4      BRCA.Normal  )"
      ]
     },
     "execution_count": 354,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cleaned_survival_df.shape, cleaned_survival_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# cleaned_sampleMap_df와 cleaned_survival_df 데이터 병합"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 355,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "병합 후 데이터 크기: (1037, 84)\n"
     ]
    }
   ],
   "source": [
    "# 공통된 환자 ID를 기준으로 데이터 병합\n",
    "cleaned_survival_df['sampleID'] = cleaned_survival_df['sample']\n",
    "cleaned_survival_df = cleaned_survival_df.drop('sample', axis=1)\n",
    "merged_data = pd.merge(cleaned_sampleMap_df, cleaned_survival_df, how='inner', on='sampleID')\n",
    "\n",
    "# 병합 결과 확인\n",
    "print(\"병합 후 데이터 크기:\", merged_data.shape)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 357,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "공통된 고유 환자 ID 수: 1037\n"
     ]
    }
   ],
   "source": [
    "# 중복 여부 확인 및 공통 ID 수 계산\n",
    "sampleMap_unique_ids = set(cleaned_sampleMap_df['sampleID'].dropna().unique())\n",
    "survival_unique_ids = set(cleaned_survival_df['sampleID'].dropna().unique())\n",
    "\n",
    "# 공통 ID 수 계산\n",
    "common_ids = sampleMap_unique_ids.intersection(survival_unique_ids)\n",
    "num_common_ids = len(common_ids)\n",
    "\n",
    "print(f\"공통된 고유 환자 ID 수: {num_common_ids}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "그러므로 cleaned_sampleMap_df 데이터와 cleaned_survival_df 데이터의 공통된 고유 환자 ID는 1037명이다."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 358,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sampleID</th>\n",
       "      <th>AJCC_Stage_nature2012</th>\n",
       "      <th>Age_at_Initial_Pathologic_Diagnosis_nature2012</th>\n",
       "      <th>Days_to_Date_of_Last_Contact_nature2012</th>\n",
       "      <th>ER_Status_nature2012</th>\n",
       "      <th>Gender_nature2012</th>\n",
       "      <th>HER2_Final_Status_nature2012</th>\n",
       "      <th>Metastasis_Coded_nature2012</th>\n",
       "      <th>Metastasis_nature2012</th>\n",
       "      <th>Node_Coded_nature2012</th>\n",
       "      <th>...</th>\n",
       "      <th>year_of_initial_pathologic_diagnosis</th>\n",
       "      <th>Subtype</th>\n",
       "      <th>OS</th>\n",
       "      <th>OS.time</th>\n",
       "      <th>DSS</th>\n",
       "      <th>DSS.time</th>\n",
       "      <th>DFI</th>\n",
       "      <th>DFI.time</th>\n",
       "      <th>PFI</th>\n",
       "      <th>PFI.time</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>TCGA-3C-AAAU-01</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>2004.0</td>\n",
       "      <td>BRCA.LumA</td>\n",
       "      <td>0</td>\n",
       "      <td>4047.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4047.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1808.0</td>\n",
       "      <td>1</td>\n",
       "      <td>1808.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>TCGA-3C-AALI-01</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>2003.0</td>\n",
       "      <td>BRCA.Her2</td>\n",
       "      <td>0</td>\n",
       "      <td>4005.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4005.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4005.0</td>\n",
       "      <td>0</td>\n",
       "      <td>4005.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>TCGA-3C-AALJ-01</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>2011.0</td>\n",
       "      <td>BRCA.LumB</td>\n",
       "      <td>0</td>\n",
       "      <td>1474.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1474.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1474.0</td>\n",
       "      <td>0</td>\n",
       "      <td>1474.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>TCGA-4H-AAAK-01</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>2013.0</td>\n",
       "      <td>BRCA.LumA</td>\n",
       "      <td>0</td>\n",
       "      <td>348.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>348.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>348.0</td>\n",
       "      <td>0</td>\n",
       "      <td>348.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>TCGA-A1-A0SB-01</td>\n",
       "      <td>Stage I</td>\n",
       "      <td>70.0</td>\n",
       "      <td>259.0</td>\n",
       "      <td>Positive</td>\n",
       "      <td>FEMALE</td>\n",
       "      <td>Negative</td>\n",
       "      <td>Negative</td>\n",
       "      <td>M0</td>\n",
       "      <td>Negative</td>\n",
       "      <td>...</td>\n",
       "      <td>2008.0</td>\n",
       "      <td>BRCA.Normal</td>\n",
       "      <td>0</td>\n",
       "      <td>259.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>259.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>259.0</td>\n",
       "      <td>0</td>\n",
       "      <td>259.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 83 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          sampleID AJCC_Stage_nature2012  \\\n",
       "0  TCGA-3C-AAAU-01                   NaN   \n",
       "1  TCGA-3C-AALI-01                   NaN   \n",
       "2  TCGA-3C-AALJ-01                   NaN   \n",
       "3  TCGA-4H-AAAK-01                   NaN   \n",
       "4  TCGA-A1-A0SB-01               Stage I   \n",
       "\n",
       "   Age_at_Initial_Pathologic_Diagnosis_nature2012  \\\n",
       "0                                             NaN   \n",
       "1                                             NaN   \n",
       "2                                             NaN   \n",
       "3                                             NaN   \n",
       "4                                            70.0   \n",
       "\n",
       "   Days_to_Date_of_Last_Contact_nature2012 ER_Status_nature2012  \\\n",
       "0                                      NaN                  NaN   \n",
       "1                                      NaN                  NaN   \n",
       "2                                      NaN                  NaN   \n",
       "3                                      NaN                  NaN   \n",
       "4                                    259.0             Positive   \n",
       "\n",
       "  Gender_nature2012 HER2_Final_Status_nature2012 Metastasis_Coded_nature2012  \\\n",
       "0               NaN                          NaN                         NaN   \n",
       "1               NaN                          NaN                         NaN   \n",
       "2               NaN                          NaN                         NaN   \n",
       "3               NaN                          NaN                         NaN   \n",
       "4            FEMALE                     Negative                    Negative   \n",
       "\n",
       "  Metastasis_nature2012 Node_Coded_nature2012  ...  \\\n",
       "0                   NaN                   NaN  ...   \n",
       "1                   NaN                   NaN  ...   \n",
       "2                   NaN                   NaN  ...   \n",
       "3                   NaN                   NaN  ...   \n",
       "4                    M0              Negative  ...   \n",
       "\n",
       "  year_of_initial_pathologic_diagnosis      Subtype  OS OS.time  DSS DSS.time  \\\n",
       "0                               2004.0    BRCA.LumA   0  4047.0  0.0   4047.0   \n",
       "1                               2003.0    BRCA.Her2   0  4005.0  0.0   4005.0   \n",
       "2                               2011.0    BRCA.LumB   0  1474.0  0.0   1474.0   \n",
       "3                               2013.0    BRCA.LumA   0   348.0  0.0    348.0   \n",
       "4                               2008.0  BRCA.Normal   0   259.0  0.0    259.0   \n",
       "\n",
       "   DFI DFI.time PFI  PFI.time  \n",
       "0  1.0   1808.0   1    1808.0  \n",
       "1  0.0   4005.0   0    4005.0  \n",
       "2  0.0   1474.0   0    1474.0  \n",
       "3  0.0    348.0   0     348.0  \n",
       "4  0.0    259.0   0     259.0  \n",
       "\n",
       "[5 rows x 83 columns]"
      ]
     },
     "execution_count": 358,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "merged_data.drop(columns=['Subtype_Selected_y'], axis=1, inplace=True)\n",
    "merged_data.rename(columns={'Subtype_Selected_x': 'Subtype'}, inplace=True)\n",
    "merged_data.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 359,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "sampleID                                            0\n",
       "AJCC_Stage_nature2012                             353\n",
       "Age_at_Initial_Pathologic_Diagnosis_nature2012    248\n",
       "Days_to_Date_of_Last_Contact_nature2012           258\n",
       "ER_Status_nature2012                              359\n",
       "                                                 ... \n",
       "DSS.time                                            0\n",
       "DFI                                                 0\n",
       "DFI.time                                            0\n",
       "PFI                                                 0\n",
       "PFI.time                                            0\n",
       "Length: 83, dtype: int64"
      ]
     },
     "execution_count": 359,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "merged_data.isnull().sum()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 363,
   "metadata": {},
   "outputs": [],
   "source": [
    "merged_data.to_csv('/Users/zoohunn/Desktop/비어플/비어플[의료]/merged_data.csv', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "python (myenv)",
   "language": "python",
   "name": "myenv"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
